packages/headless/src/app/engine.ts

```diff
@@ -0,0 +1,248 @@
+   * type Middleware = (api: MiddlewareAPI) => (next: Dispatch) => Dispatch
+   * ```
+   * [Redux documentation on middlewares.](https://redux.js.org/glossary#middleware)
+   */
+  middlewares?: Middleware<{}, StateFromReducersMapObject<Reducers>>[];
+  loggerOptions?: LoggerOptions;
+}
+
+interface SearchAPIClientArgument {
+  searchAPIClient: SearchAPIClient;
+}
+
+export function buildEngine<Reducers extends ReducersMapObject>(
+  options: EngineOptions<Reducers>,
+  thunkExtraArguments: SearchAPIClientArgument,
+  logger: Logger
+): Engine<StateFromReducersMapObject<Reducers>> {
+  const engine = buildCoreEngine(options, thunkExtraArguments, logger);
+  const {
+    accessToken,
+    organizationId,
+    platformUrl,
+    analytics,
+  } = options.configuration;
+
+  engine.dispatch(
+    updateBasicConfiguration({
+      accessToken,
+      organizationId,
+      platformUrl,
+    })
+  );
+
+  if (analytics) {
+    const {analyticsClientMiddleware, ...rest} = analytics;
+    engine.dispatch(updateAnalyticsConfiguration(rest));
+  }
+
+  return engine;
+}
+
+function buildCoreEngine<Reducers extends ReducersMapObject>(
+  options: EngineOptions<Reducers>,
+  thunkExtraArguments: SearchAPIClientArgument,
+  logger: Logger
+): Engine<StateFromReducersMapObject<Reducers>> {
+  const {configuration, reducers} = options;
+  const reducerManager = createReducerManager(reducers);
+  const store = createStore(
+    options,
+    thunkExtraArguments,
+    logger,
+    reducerManager
+  );
+
+  return {
+    renewAccessToken: createRenewAccessTokenFunction(
+      configuration,
+      store.dispatch
+    ),
+
+    addReducers(reducers: ReducersMapObject) {
+      reducerManager.add(reducers);
+      store.replaceReducer(reducerManager.combinedReducer);
+    },
+
+    dispatch: store.dispatch,
+
+    subscribe: store.subscribe,
+
+    get state() {
+      return store.getState();
+    },
+
+    logger,
+
+    store,
+  };
+}
+
+function createStore<Reducers extends ReducersMapObject>(
+  options: EngineOptions<Reducers>,
+  thunkExtraArguments: SearchAPIClientArgument,
+  logger: Logger,
+  reducerManager: ReducerManager
+) {
+  const {preloadedState, middlewares, configuration} = options;
+
+  const analyticsClientMiddleware = getAnalyticsClientMiddleware(configuration);
+  const validatePayload = validatePayloadAndThrow;
+  const preprocessRequest = getPreprocessRequest(configuration);
+
+  return configureStore({
+    preloadedState,
+    reducer: reducerManager.combinedReducer,
+    middlewares,
+    thunkExtraArguments: {
+      logger,
+      analyticsClientMiddleware,
+      validatePayload,
+      preprocessRequest,
+      ...thunkExtraArguments,
+    },
+  });
+}
+
+function getAnalyticsClientMiddleware(
+  configuration: EngineConfigurationOptions
+): AnalyticsClientSendEventHook {
+  const {analytics} = configuration;
+  return analytics?.analyticsClientMiddleware || ((_, p) => p);
```

Comment:
nit: Could do like line 215 and have a named function `NoopAnalyticsMiddleware` instead of this anonymous one (or, instead change line 215 to be an anonymous function instead, as you prefer).

Just for uniformity.

Reply:
Sounds good